saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/OpenSamlAuthenticationProvider.java

```diff
@@ -474,6 +480,33 @@ else if (subject.getEncryptedID() != null) {
 		return username;
 	}
 
+	private Map<String, Object> getAssertionAttributes(Assertion assertion) {
+		Map<String, Object> attributeMap = new HashMap<>();
+		for (AttributeStatement attributeStatement : assertion.getAttributeStatements()) {
+			for (Attribute attribute : attributeStatement.getAttributes()) {
+				List<String> attributeValues = new ArrayList<>();
+				for (XMLObject xmlObject : attribute.getAttributeValues()) {
+					attributeValues.add(getXmlObjectValue(xmlObject));
+				}
+				attributeMap.put(attribute.getName(), attributeValues);
+			}
+		}
+		return attributeMap;
+	}
+
+	private String getXmlObjectValue(XMLObject xmlObject) {
+		if (xmlObject == null) {
+			return null;
+		}
+		if (xmlObject instanceof XSString) {
+			((XSString) xmlObject).getValue();
+		}
+		if (xmlObject instanceof XSAnyImpl) {
```

Comment:
I think there are a few more types here that are relevant like `XSBoolean`, `XSDateTime`, `XSInteger`, and `XSURI`.